client/src/main/scala/org/http4s/client/websocket/WSClient.scala

```diff
@@ -0,0 +1,224 @@
+          Eval.later(x.method.compare(y.method)),
+          Eval.later(x.uri.compare(y.uri)),
+        )
+    }
+
+  implicit val catsShowForWSRequest: Show[WSRequest] =
+    Show.fromToString
+
+  implicit def stdLibOrdering: Ordering[WSRequest] =
+    catsHashAndOrderForWSRequest.toOrdering
+}
+
+sealed trait WSFrame extends Product with Serializable
+
+sealed trait WSControlFrame extends WSFrame
+
+sealed trait WSDataFrame extends WSFrame
+
+object WSFrame {
+  final case class Close(statusCode: Int, reason: String) extends WSControlFrame
+  final case class Ping(data: ByteVector) extends WSControlFrame
+  final case class Pong(data: ByteVector) extends WSControlFrame
+  final case class Text(data: String, last: Boolean = true) extends WSDataFrame
+  final case class Binary(data: ByteVector, last: Boolean = true) extends WSDataFrame
+}
+
+trait WSConnection[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Foldable, A <: WSFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits a `()` for each chunk sent. */
+  final def sendPipe: Pipe[F, WSFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Wait for a single websocket frame to be received. Returns `None` if the receiving side is
+    * closed.
+    */
+  def receive: F[Option[WSFrame]]
+
+  /** A stream of the incoming websocket frames. */
+  final def receiveStream: Stream[F, WSFrame] = Stream.repeatEval(receive).unNoneTerminate
+
+  /** The negotiated subprotocol, if any. */
+  def subprotocol: Option[String]
+}
+
+trait WSConnectionHighLevel[F[_]] {
+
+  /** Send a single websocket frame. The sending side of this connection has to be open. */
+  def send(wsf: WSDataFrame): F[Unit]
+
+  /** Send multiple websocket frames. Equivalent to multiple `send` calls, but at least as fast. */
+  def sendMany[G[_]: Foldable, A <: WSDataFrame](wsfs: G[A]): F[Unit]
+
+  /** A `Pipe` which sends websocket frames and emits a `()` for each chunk sent. */
+  final def sendPipe: Pipe[F, WSDataFrame, Unit] = _.chunks.evalMap(sendMany(_))
+
+  /** Send a Close frame. The sending side of this connection will be closed. */
+  def sendClose(reason: String = ""): F[Unit]
+
+  /** Wait for a websocket frame to be received. Returns `None` if the receiving side is closed.
+    * Fragmentation is handled automatically, the `last` attribute can be ignored.
+    */
+  def receive: F[Option[WSDataFrame]]
+
+  /** A stream of the incoming websocket frames. */
+  final def receiveStream: Stream[F, WSDataFrame] = Stream.repeatEval(receive).unNoneTerminate
+
+  /** The negotiated subprotocol, if any. */
+  def subprocotol: Option[String]
+
+  /** The close frame, if available. */
+  def closeFrame: TryableDeferred[F, WSFrame.Close]
+}
+
+trait WSClientHighLevel[F[_]] {
```

Comment:
I know the "high level" distinction is not so popular but it's an important distinction in practice between what we can support in the browser and what we can't.

Reply:
@armanbilge we might add a brief comment to both traits. I fear it may not be obvious to a casual reader why this distinction is made.